sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -137,18 +110,86 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                 SyntaxKind.ObjectCreationExpression);
         }
 
-        private bool HasOverloadThatUsesUriTypeInPlaceOfString(IMethodSymbol originalMethodSymbol, ISet<int> paramIdx)
+        private void VerifyDeclaration(SyntaxNodeAnalysisContext context)
         {
+            var methodDeclaration = (BaseMethodDeclarationSyntax)context.Node;
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+            if (methodDeclaration == null || methodSymbol == null || methodSymbol.IsOverride)
+            {
+                return;
+            }
+
+            VerifyReturnType(context, methodDeclaration, methodSymbol);
+
+            var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
+            if (!stringUrlParams.Any())
+            {
+                return;
+            }
+
+            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(methodSymbol, stringUrlParams);
+            if (methodOverloads.Any())
+            {
+                var methodOverloadSet = new HashSet<IMethodSymbol>(methodOverloads);
+                if (!methodDeclaration.IsKind(SyntaxKind.ConstructorDeclaration) &&
+                    !SyntaxHelper.ContainsMethodInvocation(methodDeclaration, context.SemanticModel,
+                        mSyntax => true,
+                        mSymbol => methodOverloadSet.Contains(mSymbol)))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3997, FindIdentifierLocation(methodDeclaration)));
+                }
+            }
+            else
+            {
+                foreach (int paramIdx in stringUrlParams)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3994,
+                        methodDeclaration.ParameterList.Parameters[paramIdx].Type.GetLocation()));
+                }
+            }
+        }
+
+        private Location FindIdentifierLocation(BaseMethodDeclarationSyntax baseMethodDeclaration)
```

Comment:
I don't like this one, also because BaseMethodDeclarationSyntax could be something else than these two if we register another SyntaxKind. Suggestions welcome.